nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,164 @@
+
+  // Forward integrate robot position and orientation using velocity commands
+  tf2::Matrix3x3 rotation_matrix(base_link_orientation);
+  tf2::Vector3 linear_velocity(msg->linear.x, msg->linear.y, msg->linear.z);
+  tf2::Vector3 rotated_linear_velocity = rotation_matrix * linear_velocity;
+
+  odom_updated_pose_.pose.pose.position.x += rotated_linear_velocity.x() * dt;
+  odom_updated_pose_.pose.pose.position.y += rotated_linear_velocity.y() * dt;
+  odom_updated_pose_.pose.pose.position.z += rotated_linear_velocity.z() * dt;
+
+  // Broadcast "odom" to "base_link" transform using the local variable
+  geometry_msgs::msg::TransformStamped odom_to_base_transform;
+  odom_to_base_transform.header.stamp = odom_updated_pose_.header.stamp;
+  odom_to_base_transform.header.frame_id = "odom";
+  odom_to_base_transform.child_frame_id = "base_link";
+  odom_to_base_transform.transform.translation.x = odom_updated_pose_.pose.pose.position.x;
+  odom_to_base_transform.transform.translation.y = odom_updated_pose_.pose.pose.position.y;
+  odom_to_base_transform.transform.translation.z = odom_updated_pose_.pose.pose.position.z;
+  odom_to_base_transform.transform.rotation = odom_updated_pose_.pose.pose.orientation;
+
+  tf_broadcaster_->sendTransform(odom_to_base_transform);
+}
+
+void LoopbackSimulator::initposeCallback(
+  const
+  geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  RCLCPP_INFO(this->get_logger(), "Received initial pose wrt map");
+  init_pose_ = *msg;
+  init_pose_set_ = true;
+
+  if (init_odom_base_published_ == false) {
+    // Publish Identity transform between odom to base_footprint
+    geometry_msgs::msg::TransformStamped odom_to_base_transform;
+    odom_to_base_transform.header.stamp = this->now();
+    odom_to_base_transform.header.frame_id = "odom";
+    odom_to_base_transform.child_frame_id = "base_footprint";
+    odom_to_base_transform.transform.translation.x = 0.0;
+    odom_to_base_transform.transform.translation.y = 0.0;
+    odom_to_base_transform.transform.translation.z = 0.0;
+    tf2::Quaternion q;
+    q.setRPY(0, 0, 0);
+    odom_to_base_transform.transform.rotation.x = q.x();
+    odom_to_base_transform.transform.rotation.y = q.y();
+    odom_to_base_transform.transform.rotation.z = q.z();
+    odom_to_base_transform.transform.rotation.w = q.w();
+
+    tf_broadcaster_->sendTransform(odom_to_base_transform);
+    init_odom_base_published_ = true;
+  }
+}
+
+void LoopbackSimulator::timerCallback()
+{
+  if (init_pose_set_ == false) {return;}
+  // RCLCPP_INFO(this->get_logger(), "Publishing map->odom tf");
+
+  // Publish map to odom transform
+  geometry_msgs::msg::TransformStamped map_to_odom_transform;
+  map_to_odom_transform.header.stamp = this->now();
+  map_to_odom_transform.header.frame_id = "map";
+  map_to_odom_transform.child_frame_id = "odom";
+  map_to_odom_transform.transform.translation.x = init_pose_.pose.pose.position.x;
```

Comment:
This gets executed only when an initial pose estimate is given from Rviz, that means a 0-transform is also published between "odom" and "base_link". 

So it makes more sense to me to continously publish the above transform between map->odom, because odom is a fixed frame

Reply:
OK